File: services/server/src/safe_socket/safe_socket.py
import socket


def recv_all(socket: socket.socket, size):
    bytesAmount = 0
    readBytes = []
    while bytesAmount < size:
        bytes = socket.recv(size - bytesAmount)
        if not bytes:
            return b""
        readBytes.append(bytes)
        bytesAmount += len(bytes)
    return b"".join(readBytes)


def send_all(socket: socket.socket, bytes):
    wroteBytes = 0
    while wroteBytes < len(bytes):
        n = socket.send(bytes[wroteBytes:])
        wroteBytes += n
    return wroteBytes

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

//TODO: Complete with a short-read/short-write tolerant implementation

func SendAll(socket io.Writer, bytes []byte) error {
	wroteBytes := 0
	for wroteBytes < len(bytes) {
		n, err := socket.Write(bytes[wroteBytes:])
		if err != nil {
			return err
		}
		wroteBytes += n
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	readBytes := 0
	for readBytes < size {
		n, err := socket.Read(buff[readBytes:])
		if err != nil {
			return nil, err
		}
		readBytes += n
	}
	return buff, nil
}
